VultisigApp/VultisigApp/Views/Vault/CustomTokenView.swift

```diff
@@ -0,0 +1,257 @@
+        VStack(alignment: .leading, spacing: 8) {
+            Text(NSLocalizedString(title, comment: ""))
+                .font(.body20MontserratSemiBold)
+                .foregroundColor(.neutral0)
+            
+            Text(address)
+                .font(.body12Menlo)
+                .foregroundColor(.turquoise600)
+        }
+        .frame(maxWidth: .infinity, alignment: .leading)
+    }
+    
+    private func getDetailsCell(for title: String, with value: String) -> some View {
+        HStack {
+            Text(NSLocalizedString(title, comment: ""))
+            Spacer()
+            Text(value)
+        }
+        .font(.body16MenloBold)
+        .foregroundColor(.neutral100)
+    }
+    
+    func errorView(error: Error) -> some View {
+        return VStack(spacing: 16) {
+            Text(error.localizedDescription)
+                .multilineTextAlignment(.center)
+                .font(.body16Menlo)
+                .foregroundColor(.neutral0)
+                .padding(.horizontal, 16)
+            
+            Button {
+                Task { await fetchTokenInfo() }
+            } label: {
+                FilledButton(title: "Retry")
+            }
+            .padding(.horizontal, 40)
+        }
+        .frame(maxWidth: .infinity, maxHeight: .infinity)
+    }
+    
+    private func fetchTokenInfo() async {
+        guard !contractAddress.isEmpty else { return }
+        isLoading = true
+        showTokenInfo = false
+        error = nil
+        
+        do {
+            let service = try EvmServiceFactory.getService(forChain: group.chain)
+            let (name, symbol, decimals) = try await service.getTokenInfo(contractAddress: contractAddress)
+            
+            if !name.isEmpty, !symbol.isEmpty, decimals > 0 {
+                
+                let nativeTokenOptional = group.coins.first(where: {$0.isNativeToken})
+                if let nativeToken = nativeTokenOptional {
+                    self.token = Coin(
+                        chain: nativeToken.chain,
+                        ticker: symbol,
+                        logo: .empty,
+                        address: nativeToken.address,
+                        priceRate: .zero,
+                        decimals: decimals,
+                        hexPublicKey: nativeToken.hexPublicKey,
+                        priceProviderId: .empty,
+                        contractAddress: contractAddress,
+                        rawBalance: .zero,
+                        isNativeToken: false
+                    )
+                    
+                    if let customToken = self.token {
+                        let (rawBalance, priceRate) = try await service.getBalance(coin: customToken)
+                        self.token?.rawBalance = rawBalance
+                        self.token?.priceRate = priceRate
+                        
+                        DispatchQueue.main.async {
+                            self.tokenName = name
```

Comment:
we have async context here, you can use @MainActor

Reply:
Fixed